File: Bee.m

classdef Bee < handle
    properties
        ns
        nb
        ne
        nrb
        nre
        iterations
        bees
        best_objectives
        animation
        problem 
    end
    methods
        function obj = Bee(ns, nb, ne, nrb, nre, iterations, animation, problem)
            obj.ns = ns;
            obj.nb = nb;
            obj.ne = ne;
            obj.nrb = nrb;
            obj.nre = nre;
            obj.iterations = iterations;
            obj.bees = [];
            obj.problem = problem;
            obj.animation = animation;
        end
        function solve(obj)
            fig = figure;
            tic;
            obj.initialRandSolution();
            for t = 1:obj.iterations
                obj.eliteSearch();
                obj.bestSearch();
                obj.globalFill();
                obj.calculateBests();
                if obj.animation
                    obj.problem.animate(fig, t, obj.bees(1, 1:end-1), obj.bees(1, end), obj.best_objectives);
                end
            end
            elapsedTime = toc; 
            fprintf('Iterations: %.0f\n', obj.iterations)
            fprintf('Total elapsed time: %.2f seconds\n', elapsedTime);
            best_solution = obj.bees(1, 1:end-1);
            fprintf('Best solution: %s\n', mat2str(best_solution));
            best_objective = obj.bees(1, end);
            fprintf('Best objective: %.3f\n', best_objective);
            if ~ obj.animation
                obj.problem.visualize(best_solution, best_objective);
            end
        end
        function eliteSearch(obj)
            for e = 1:obj.ne
                bestValue = inf;
                bestMutation = [];
                for i = 1:obj.nre
                    mutation = obj.problem.mutate(obj.bees(e, :));
                    mutationObjective = obj.problem.eval(mutation);
                    if mutationObjective < bestValue
                        bestMutation = mutation;
                        bestValue = mutationObjective;
                    end
                end

                if bestValue < obj.bees(e, size(obj.bees, 2))
                    obj.bees(e, :) = [bestMutation, bestValue];
                end
            end
        end
        function bestSearch(obj)
            for b = obj.ne + 2:obj.nb + 1
                bestValue = inf;
                bestMutation = [];
                for i = 1:obj.nrb
                    mutation = obj.problem.mutate(obj.bees(b, :));
                    mutationObjective = obj.problem.eval(mutation);
                    if mutationObjective < bestValue
                        bestMutation = mutation;
                        bestValue = mutationObjective;
                    end
                end
                if bestValue < obj.bees(b, size(obj.bees, 2))
                    obj.bees(b, :) = [bestMutation, bestValue];
                end
            end
        end
        function globalFill(obj)
            for g = obj.nb + 2:length(obj.bees)
                obj.bees(g, :) = [obj.problem.random(), obj.problem.eval(obj.problem.random())];
            end
        end
        function calculateBests(obj)
            obj.bees = sortrows(obj.bees, size(obj.bees, 2));
            obj.best_objectives= [obj.best_objectives; obj.bees(1, end)];
        end
        function initialRandSolution(obj)
            for iter = 1:obj.ns
                x = obj.problem.random();
                y = obj.problem.eval(x);
                obj.bees = [obj.bees; [x, y]];
            end
            obj.bees = sortrows(obj.bees, size(obj.bees, 2));
        end
    end
end

File: C++/Bee.cpp
//
// Created by Nicolas Schneider on 18.12.23.
//
#include <vector>
#include <algorithm>
#include <iostream>
#include <numeric>
#include <random>
#include <cmath>
#include <chrono>
#include "TSP.h"
#include "Bee.h"

Bee::Bee(int ns, int nb, int ne, int nrb, int nre, int iterations, TSP problem)
    : ns(ns), nb(nb), ne(ne), nrb(nrb), nre(nre), iterations(iterations), problem(problem) {
    this->coords = problem.coords;
    this->distances = problem.distances;

    // Initialize best_objectives
    this->best_objectives = std::vector<double>(iterations, std::numeric_limits<double>::infinity());

    // Initialize bees
    this->bees = std::vector<std::vector<double>>(ns, std::vector<double>(problem.routeLen, 1.0));

}

void Bee::solve() {
    auto start = std::chrono::high_resolution_clock::now();
    initialRandSolution();
    double lastBestValue = std::numeric_limits<double>::infinity();
    int noImprovementCounter = 0;
    int noImprovementThreshold = 100;  // Set this to the number of iterations without improvement you tolerate
    bool improvementThresholdReached = false;

    for (int t = 0; t < iterations; ++t) {
        eliteSearch();
        bestSearch();
        globalFill();
        calculateBests();

        // Check for improvement
        if (bees[0].back() < lastBestValue) {
            lastBestValue = bees[0].back();
            noImprovementCounter = 0;
        } else {
            noImprovementCounter++;
        }

        // Check if we have reached the no improvement threshold
        if (noImprovementCounter >= noImprovementThreshold) {
            std::cout << "Terminating after " << noImprovementThreshold << " iterations without improvement." << std::endl;
            improvementThresholdReached = true;
            break;
        }
    }

    std::vector<double> best_solution = bees[0];
    best_solution.pop_back();
    if (!improvementThresholdReached) {
        std::cout << "Terminating after " << iterations << " iterations." << std::endl;
    }
    std::cout << "\n----------------------\n";

    std::cout << "Route: ";
    for (auto &val: best_solution) {
        std::cout << val << " ";
    }
    std::cout << "\nObjective Value: " << bees[0].back() << std::endl;

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = end - start;
    auto duration_sec = std::chrono::duration_cast<std::chrono::seconds>(duration).count();
    auto duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(duration).count() - 1000 * duration_sec;

    // Print total time in seconds and milliseconds
    std::cout << "Total Time: " << duration_sec << "s " << duration_ms << "ms" << std::endl;
    std::cout << "\n========FINISHED========\n";
    
}

void Bee::initialRandSolution() {
    for (int iter = 0; iter < ns; ++iter) {
        std::vector<double> x = problem.random();
        x.push_back(problem.eval(x));
        bees[iter] = x;
    }
    std::sort(bees.begin(), bees.end(), [](const std::vector<double>& a, const std::vector<double>& b) {
        return a.back() < b.back();
    });
}

void Bee::calculateBests() {
    std::sort(bees.begin(), bees.end(), [](const std::vector<double>& a, const std::vector<double>& b) {
        return a.back() < b.back();
    });

    best_objectives.push_back(bees[0].back());
}

void Bee::eliteSearch() {
    for (int e = 0; e < ne; ++e) {
        double bestValue = std::numeric_limits<double>::infinity();
        std::vector<double> bestMutation;
        for (int i = 0; i < nre; ++i) {
            // print bees[e]

            std::vector<double> beeWithoutLastElement = std::vector<double>(bees[e].begin(), bees[e].end() - 1);
            std::vector<double> mutation = problem.mutate(beeWithoutLastElement);
            double mutationObjective = problem.eval(mutation);
            if (mutationObjective < bestValue) {
                bestMutation = mutation;
                bestValue = mutationObjective;
            }
        }
        if (bestValue < bees[e].back()) {
            bestMutation.push_back(bestValue);
            bees[e] = bestMutation;
        }
    }
}

void Bee::bestSearch() {
    for (int b = 0; b < nb; ++b) {
        double bestValue = std::numeric_limits<double>::infinity();
        std::vector<double> bestMutation;
        for (int i = 0; i < nrb; ++i) {
            std::vector<double> beeWithoutLastElement = std::vector<double>(bees[b].begin(), bees[b].end() - 1);
            std::vector<double> mutation = problem.mutate(beeWithoutLastElement);
            double mutationObjective = problem.eval(mutation);
            if (mutationObjective < bestValue) {
                bestMutation = mutation;
                bestValue = mutationObjective;
            }
        }
        if (bestValue < bees[b].back()) {
            bestMutation.push_back(bestValue);
            bees[b] = bestMutation;
        }
    }
}

void Bee::globalFill() {
    for (int g = nb + 1; g < bees.size(); ++g) {
        std::vector<double> randomSolution = problem.random();
        randomSolution.push_back(problem.eval(randomSolution));
        bees[g] = randomSolution;
    }
}






